File: leetcode/42.js
// Given n non-negative integers representing an elevation map where the width
// of each bar is 1, compute how much water it is able to trap after raining.
//
// Example:
//
// Input: [0,1,0,2,1,0,1,3,2,1,2,1]
// Output: 6

const trap = heights => {
  let left_max = new Array(heights.length);
  let right_max = new Array(heights.length);
  let water = 0;
  // to avoid calculating the left and right max separately for each index
  // we need to store the result of previous examinations and then compare
  // it iteratively to the current index. because its a maximum, it follows
  // that the next index can use the results of examining all the previous
  // indexes for local maximums

  // we seed the first index with the height as the first maximum, because
  // the edges of the array are "emtpy" and therefor contain no maximum to
  // compare to
  left_max[0] = heights[0];

  // start at 1, as 0 is seeded
  for (let i = 1; i < heights.length; i++) {
    // calculate each indexes left_max based on the left_max of the previous
    // index. we compare it to the height of the current index in case we
    // find a new maximum at our current index
    left_max[i] = Math.max(heights[i], left_max[i - 1]);
  }

  // similar to left we have to seed the right
  right_max[heights.length - 1] = heights[heights.length - 1];

  // this has to go right to left here, as our optimization depends on using
  // data from the edges of the array inwards due to the nature of the
  // problem (e.g. water is trapped by creating a cavity)
  for (let i = heights.length - 2; i >= 0; i--) {
    right_max[i] = Math.max(heights[i], right_max[i + 1]);
  }

  // lastly loop through each index and find the minimum of the local maximums
  // for both left and right to find the "ceiling" for the water trapped
  // at each index
  for (let i = 0; i < heights.length; i++) {
    water += Math.min(left_max[i], right_max[i]) - heights[i];
  }

  return water;
};

console.log(trap([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]));

File: season2/42.py
class Solution:
    def trap(self, heights: List[int]) -> int:
        # at each i, check the left and right max to see if its higher than
        # the current point. if it is, this index can trap water
        # this is a bit of a dynamic programming problem because the left/right
        # max can be calculated based on the left/right max of the previous index
        
        left_max = [0] * len(heights)
        right_max = [0] * len(heights)
        
        
        for rev_i, height in enumerate(reversed(heights)):
            i = len(heights) - rev_i - 1
            right_max[i] = max(right_max[i+1] if i + 1 < len(heights) else 0, height)
        
        trapped = 0
            
        for i, height in enumerate(heights):
            #print(i, height, left_max[i-1] if i - 1 >= 0 else 0)
            left_max[i] = max(left_max[i-1] if i - 1 >= 0 else 0, height)
            
            if height < max(left_max[i], right_max[i]):
                trapped += min(left_max[i], right_max[i]) - height
        
        #print(left_max, right_max)
        
        return trapped
        

